### DISCUSSION TITLE
New GoPro Labs Release for HERO11 v2.30.70

### DISCUSSION THREAD

**Discussion by dnewman-gpsw** [gopro/labs#810] [Announcements]
Release notes below. The release is for more scripting power under the hood, and some catch-up and sync with Labs on HERO12.

### 2.30.70 - Mar 14, 2024
- Added *wxyz, permanent storage, is the simplified version of !Mwxyz, e.g. *BITR=160
- Added $wxyz, temporary storage, is the simplified version of oMwxyz, e.g. $WIDE=1$NR01=1
- Added support for macros, or crude scripting subroutines. This expands the potential complexity for QR scripting, as the entire script is no longer limited to 255 characters. It would be possible to run scripts nearing ~1.2KB. 
Note: recursion is supported, macros can call other macros, and can call themselves.
  * Subroutines are saved in string FourCCs. e.g. ```*SUBA="mVr4p60'60p'!S!2E!1N"```   Note: Use single quotes for text within double quotes.
  * Subroutines are launched ```^SUBA```  e.g. ```>19:30^NIGH~^DAYS``` (if(time>19:30) Night(); else Day(); ).
- Added support testing current camera mode, system variable 'v' will have the mode like follows.
  * video=12
  * TLV=13
  * Looping Video=15
  * Photo=16
  * Night Photo=18
  * Burst Photo=19, etc.<br>
Display the current mode with with this command:  ```"mode $v"!R``` or fast/cleaner ```$PMSG="mode $v"!R```
- Added precise variable testing with '=='. Previously only < for less_than and > for greater_than_equal were supported.
   e.g. command:   ```>10:10<11:00"do thing between 10:10 and 10:59"```
   now you can do: ```==10:10"do thing only at 10:10"```
   useful for mode testing: ```==v12"video mode, do thing only on video mode"```
- Added $GRAB=x screen grab x-frames, !F (grab now) and !F3 (grab next 3 Labs overlays) - frame grab the UI. Useful for education and bug reporting
- Added font color for script messaging. e.g. try this QR Code "\1H\2e\3l\4l\5o \6W\7o\8r\1l\2d"
- Added $STDL=x - Start Delay (default 2000ms, wait before any other commands.) $STDL=100 for less wait. Experimental scripting performance.
- Improved the font rendering (higher res fonts)
- Improved upload reliability with an optional maximum transfer time (timeout). e.g. !U3 <- upload now, for up to 3 minutes.
- Improved processing speed for looping scripts about 10-15x faster.
- Fixed ```x++``` and ```x--``` EV up and down controls (handy for scuba)
- minor Labs bug fixes

**Comment by Verkhushin** [comment]
Hello.  After the update, the motion sensor began to work inadequately, poorly, with delays, and producing random numbers.
![Screenshot_20240321_090033](https://github.com/gopro/labs/assets/164307649/d07d5eab-7f8f-4846-bec3-98efd8688d84)

There is also a problem that was in the previous firmware version.  If you turn off the display once with the oDO command, you can turn on the display by touching the screen.  But if you simultaneously print text on the screen, a problem arises.  The display flashes synchronously with the text output, or does not turn on when touched.

**Reply by dnewman-gpsw** [reply]
Can you produce all needed to reproduce this easily.

**Reply by Verkhushin** [reply]
![Motion-3_21_2024, 09_44_49](https://github.com/gopro/labs/assets/164307649/c10ff318-6e10-4df8-84e0-902d30635b36)

![oDO print-3_21_2024, 09_48_23](https://github.com/gopro/labs/assets/164307649/36715dbc-4d04-47e2-aca4-d54a3ee0b8af)

**Comment by Verkhushin** [comment]
This is the problem.  Start recording.  Everything will be fine.  You can turn on the display by tapping it.  But after executing the oDO command and stopping recording, a problem will arise.
![Screenshot_20240321_091356](https://github.com/gopro/labs/assets/164307649/4f3230aa-d5c0-4659-bec4-b8f5bfb91953)

**Comment by Fedess1986** [comment]
Hello! Will there be a labs v2.32 firmware version for Hero11 Black camera? If so is there a release date? thanks!

**Reply by dnewman-gpsw** [reply]
With QA now, should be out very soon.

**Reply by Anammox** [reply]
any chance it will include a fix to live stream (RTMP) disconnect via mobile hotspot when cell coverage is intermittent?

**Reply by dnewman-gpsw** [reply]
Not unless that was a fix included in 2.32.00, which I highly expect it was not.

**Reply by Anammox** [reply]
bummer! I posted a feature request. It's the crux of my budget IRL streaming setup.

**Reply by dnewman-gpsw** [reply]
On Labs, for the main camera firmware? This a main camera feature, Labs does very little with network anything.
